src/connection.c

```diff
@@ -115,103 +109,50 @@ void DelayedConnection_cleanup(Trigger *trigger) {
   if (trigger->is_present) {
     LF_DEBUG(CONN, "Delayed connection %p had a present value this tag. Pop it", trigger);
     trigger->is_present = false;
-    int ret = self->trigger_data_queue.pop(&self->trigger_data_queue);
-    validaten(ret);
   }
 
-  if (self->trigger_data_queue.staged) {
+  if (self->staged_payload_ptr) {
     LF_DEBUG(CONN, "Delayed connection %p had a staged value. Schedule it", trigger);
     Environment *env = self->super.super.parent->env;
     Scheduler *sched = &env->scheduler;
 
-    tag_t tag = lf_delay_tag(sched->current_tag, self->delay);
-    self->trigger_data_queue.push(&self->trigger_data_queue);
-    sched->schedule_at(sched, trigger, tag);
+    tag_t base_tag = ZERO_TAG;
+    if (self->is_physical) {
+      base_tag.time = env->get_physical_time(env);
+    } else {
+      base_tag = sched->current_tag;
+    }
+    Event event = EVENT_INIT(lf_delay_tag(base_tag, self->delay), trigger, self->staged_payload_ptr);
+    sched->schedule_at(sched, &event);
+    self->staged_payload_ptr = NULL;
   }
 }
 
-/**
- * @brief Attempts to trigger downstreams are stopped when we arrive at a delayed connection.
- * Instead we stage the value for scheduling at the end of the current tag.
- */
 void DelayedConnection_trigger_downstreams(Connection *_self, const void *value, size_t value_size) {
-  (void)value_size;
-  LF_DEBUG(CONN, "Triggering downstreams on delayed connection %p. Stage the value for later scheduling", _self);
   DelayedConnection *self = (DelayedConnection *)_self;
+  assert(value);
+  assert(value_size > 0);
+  lf_ret_t ret;
+  LF_DEBUG(CONN, "Triggering downstreams on delayed connection %p. Stage the value for later scheduling", _self);
+  Trigger *trigger = &_self->super;
   Scheduler *sched = &_self->super.parent->env->scheduler;
-
-  if (value) {
-    self->trigger_data_queue.stage(&self->trigger_data_queue, value);
-  } else {
-    throw("DelayedConnection with untyped value");
+  EventPayloadPool *pool = trigger->payload_pool;
+  if (self->staged_payload_ptr == NULL) {
+    ret = pool->allocate(pool, &self->staged_payload_ptr);
```

Comment:
so this implicitly manipulates the pointer :(

Reply:
I am not sure what you mean by "implicitly". But for DelayedConnections we must buffer the proposed event until the end of the tag and we can send it out. This is done by having the field `staged_payload_ptr` point to this payload. It can be overwritten by other calls to `lf_set` and at the end of the tag it is sent and `staged_payload_ptr` is reset to point to NULL.